// @aragon/cli/dist/commands/dao_cmds/acl_cmds/view.js
"use strict";

var _aragonjsWrapper = _interopRequireDefault(require("../utils/aragonjs-wrapper"));

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _objectSpread(target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i] != null ? arguments[i] : {}; var ownKeys = Object.keys(source); if (typeof Object.getOwnPropertySymbols === 'function') { ownKeys = ownKeys.concat(Object.getOwnPropertySymbols(source).filter(function (sym) { return Object.getOwnPropertyDescriptor(source, sym).enumerable; })); } ownKeys.forEach(function (key) { _defineProperty(target, key, source[key]); }); } return target; }

function _defineProperty(obj, key, value) { if (key in obj) { Object.defineProperty(obj, key, { value: value, enumerable: true, configurable: true, writable: true }); } else { obj[key] = value; } return obj; }

function asyncGeneratorStep(gen, resolve, reject, _next, _throw, key, arg) { try { var info = gen[key](arg); var value = info.value; } catch (error) { reject(error); return; } if (info.done) { resolve(value); } else { Promise.resolve(value).then(_next, _throw); } }

function _asyncToGenerator(fn) { return function () { var self = this, args = arguments; return new Promise(function (resolve, reject) { var gen = fn.apply(self, args); function _next(value) { asyncGeneratorStep(gen, resolve, reject, _next, _throw, "next", value); } function _throw(err) { asyncGeneratorStep(gen, resolve, reject, _next, _throw, "throw", err); } _next(undefined); }); }; }

var chalk = require('chalk');

var TaskList = require('listr');

var daoArg = require('../utils/daoArg');

var _require = require('../utils/knownApps'),
    listApps = _require.listApps;

var _require2 = require('./utils/knownRoles'),
    rolesForApps = _require2.rolesForApps;

var _require3 = require('../../../helpers/web3-fallback'),
    ensureWeb3 = _require3.ensureWeb3;

var listrOpts = require('../../../helpers/listr-options');

var Table = require('cli-table');

var knownRoles = rolesForApps();
var ANY_ENTITY = '0xFFfFfFffFFfffFFfFFfFFFFFffFFFffffFfFFFfF';
var ANY_ENTITY_TEXT = 'Any entity';
var ZERO_ADDRESS = '0x0000000000000000000000000000000000000000';
var knownApps;
exports.command = 'view <dao>';
exports.describe = 'Inspect permissions in a DAO';

exports.builder = function (yargs) {
  return daoArg(yargs);
};

var printAppName = function printAppName(appId, addr) {
  if (addr === ANY_ENTITY) return ANY_ENTITY_TEXT;
  return knownApps[appId] ? "".concat(knownApps[appId].split('.')[0], " (").concat(addr.slice(0, 6), ")") : addr.slice(0, 16) + '...';
};

var appFromProxyAddress = function appFromProxyAddress(proxyAddress, apps) {
  return apps.filter(function (app) {
    return app.proxyAddress === proxyAddress;
  })[0] || {};
};

var formatRow = function formatRow(_ref, apps) {
  var to = _ref.to,
      role = _ref.role,
      allowedEntities = _ref.allowedEntities,
      manager = _ref.manager;

  if (manager === ZERO_ADDRESS && (!allowedEntities || allowedEntities.length === 0)) {
    return null;
  }

  var formattedTo = printAppName(appFromProxyAddress(to, apps).appId, to);
  var formattedRole = knownRoles[role] || "".concat(role.slice(0, 8), "..").concat(role.slice(-6));
  if (formattedRole['id']) formattedRole = formattedRole['id'];
  var formattedAllowed = allowedEntities.reduce(function (acc, addr) {
    var allowedName = printAppName(appFromProxyAddress(addr, apps).appId, addr);
    var allowedEmoji = allowedName === ANY_ENTITY_TEXT ? '🆓' : '✅';
    return acc + '\n' + allowedEmoji + '  ' + allowedName;
  }, '').slice(1); // remove first newline

  var formattedManager = printAppName(appFromProxyAddress(manager, apps).appId, manager);
  return [formattedTo, formattedRole, formattedAllowed, formattedManager];
};

exports.handler =
/*#__PURE__*/
function () {
  var _ref3 = _asyncToGenerator(
  /*#__PURE__*/
  regeneratorRuntime.mark(function _callee(_ref2) {
    var reporter, dao, network, apm, module, silent, debug, web3, tasks;
    return regeneratorRuntime.wrap(function _callee$(_context) {
      while (1) {
        switch (_context.prev = _context.next) {
          case 0:
            reporter = _ref2.reporter, dao = _ref2.dao, network = _ref2.network, apm = _ref2.apm, module = _ref2.module, silent = _ref2.silent, debug = _ref2.debug;
            knownApps = listApps([module.appName]);
            _context.next = 4;
            return ensureWeb3(network);

          case 4:
            web3 = _context.sent;
            tasks = new TaskList([{
              title: 'Inspecting DAO Permissions',
              task: function task(ctx, _task) {
                _task.output = "Fetching permissions for ".concat(dao, "...");
                return new Promise(function (resolve, reject) {
                  var resolveIfReady = function resolveIfReady() {
                    if (ctx.acl && ctx.apps) {
                      resolve();
                    }
                  };

                  (0, _aragonjsWrapper.default)(dao, apm['ens-registry'], {
                    provider: web3.currentProvider,
                    onPermissions: function onPermissions(permissions) {
                      ctx.acl = permissions;
                      resolveIfReady();
                    },
                    onApps: function onApps(apps) {
                      ctx.apps = apps;
                      resolveIfReady();
                    },
                    onDaoAddress: function onDaoAddress(addr) {
                      ctx.daoAddress = addr;
                    },
                    onError: function onError(err) {
                      return reject(err);
                    }
                  }).catch(function (err) {
                    reporter.error('Error inspecting DAO');
                    reporter.debug(err);
                    process.exit(1);
                  });
                });
              }
            }], listrOpts(silent, debug));
            return _context.abrupt("return", tasks.run().then(function (ctx) {
              reporter.success("Successfully fetched DAO apps for ".concat(ctx.daoAddress));
              var acl = ctx.acl; // filter according to cli params will happen here

              var table = new Table({
                head: ['App', 'Action', 'Allowed entities', 'Manager'].map(function (x) {
                  return chalk.white(x);
                })
              });
              var tos = Object.keys(acl);
              var flattenedACL = tos.reduce(function (acc, to) {
                var roles = Object.keys(acl[to]);
                var permissions = roles.map(function (role) {
                  return _objectSpread({
                    to: to,
                    role: role
                  }, acl[to][role]);
                });
                return acc.concat(permissions);
              }, []);
              flattenedACL.map(function (row) {
                return formatRow(row, ctx.apps);
              }).filter(function (row) {
                return row;
              }).forEach(function (row) {
                return table.push(row);
              });
              console.log(table.toString());
              process.exit(); // force exit, as aragonjs hangs
            }));

          case 7:
          case "end":
            return _context.stop();
        }
      }
    }, _callee, this);
  }));

  return function (_x) {
    return _ref3.apply(this, arguments);
  };
}();
