// @aragon/cli/dist/helpers/solidity-extractor.js
"use strict";

function asyncGeneratorStep(gen, resolve, reject, _next, _throw, key, arg) { try { var info = gen[key](arg); var value = info.value; } catch (error) { reject(error); return; } if (info.done) { resolve(value); } else { Promise.resolve(value).then(_next, _throw); } }

function _asyncToGenerator(fn) { return function () { var self = this, args = arguments; return new Promise(function (resolve, reject) { var gen = fn.apply(self, args); function _next(value) { asyncGeneratorStep(gen, resolve, reject, _next, _throw, "next", value); } function _throw(err) { asyncGeneratorStep(gen, resolve, reject, _next, _throw, "throw", err); } _next(undefined); }); }; }

function _slicedToArray(arr, i) { return _arrayWithHoles(arr) || _iterableToArrayLimit(arr, i) || _nonIterableRest(); }

function _nonIterableRest() { throw new TypeError("Invalid attempt to destructure non-iterable instance"); }

function _iterableToArrayLimit(arr, i) { var _arr = []; var _n = true; var _d = false; var _e = undefined; try { for (var _i = arr[Symbol.iterator](), _s; !(_n = (_s = _i.next()).done); _n = true) { _arr.push(_s.value); if (i && _arr.length === i) break; } } catch (err) { _d = true; _e = err; } finally { try { if (!_n && _i["return"] != null) _i["return"](); } finally { if (_d) throw _e; } } return _arr; }

function _arrayWithHoles(arr) { if (Array.isArray(arr)) return arr; }

var fs = require('fs');

var _require = require('util'),
    promisify = _require.promisify;

var readFile = promisify(fs.readFile);

var modifiesStateAndIsPublic = function modifiesStateAndIsPublic(declaration) {
  return !declaration.match(/\b(internal|private|view|pure|constant)\b/);
};

var typeOrAddress = function typeOrAddress(type) {
  var types = ['address', 'byte', 'uint', 'int', 'bool', 'string']; // check if the type starts with any of the above types, otherwise it is probably
  // a typed contract, so we need to return address for the signature

  return types.filter(function (t) {
    return type.indexOf(t) === 0;
  }).length > 0 ? type : 'address';
}; // extracts function signature from function declaration


var getSignature = function getSignature(declaration) {
  var _declaration$match$1$ = declaration.match(/function ([^]*?)\)/)[1].split('('),
      _declaration$match$1$2 = _slicedToArray(_declaration$match$1$, 2),
      name = _declaration$match$1$2[0],
      params = _declaration$match$1$2[1];

  if (!name) {
    return 'fallback';
  }

  if (params) {
    // Has parameters
    params = params.replace(/\n/gm, '').replace(/\t/gm, '').split(',').map(function (param) {
      return param.split(' ').filter(function (s) {
        return s.length > 0;
      })[0];
    }).map(function (type) {
      return typeOrAddress(type);
    }).join(',');
  }

  return "".concat(name, "(").concat(params, ")");
};

var getNotice = function getNotice(declaration) {
  // capture from @notice to either next '* @' or end of comment '*/'
  var notices = declaration.match(/(@notice)([^]*?)(\* @|\*\/)/m);
  if (!notices || notices.length === 0) return null;
  return notices[0].replace('*/', '').replace('* @', '').replace('@notice ', '').replace(/\n/gm, '').replace(/\t/gm, '').split(' ').filter(function (x) {
    return x.length > 0;
  }).join(' ');
}; // extracts required role from function declaration


var getRoles = function getRoles(declaration) {
  var auths = declaration.match(/auth.?\(([^]*?)\)/gm);
  if (!auths) return [];
  return auths.map(function (authStatement) {
    return authStatement.split('(')[1].split(',')[0].split(')')[0];
  });
}; // Takes the path to a solidity file and extracts public function signatures,
// its auth role if any and its notice statement


module.exports =
/*#__PURE__*/
function () {
  var _ref = _asyncToGenerator(
  /*#__PURE__*/
  regeneratorRuntime.mark(function _callee(sourceCodePath) {
    var sourceCode, funcDecs;
    return regeneratorRuntime.wrap(function _callee$(_context) {
      while (1) {
        switch (_context.prev = _context.next) {
          case 0:
            _context.next = 2;
            return readFile(sourceCodePath, 'utf8');

          case 2:
            sourceCode = _context.sent;
            // everything between every 'function' and '{' and its @notice
            funcDecs = sourceCode.match(/(@notice|^\s*function)(?:[^]*?){/gm);

            if (funcDecs) {
              _context.next = 6;
              break;
            }

            return _context.abrupt("return", []);

          case 6:
            return _context.abrupt("return", funcDecs.filter(function (dec) {
              return modifiesStateAndIsPublic(dec);
            }).map(function (dec) {
              return {
                sig: getSignature(dec),
                roles: getRoles(dec),
                notice: getNotice(dec)
              };
            }));

          case 7:
          case "end":
            return _context.stop();
        }
      }
    }, _callee, this);
  }));

  return function (_x) {
    return _ref.apply(this, arguments);
  };
}();
